src/Features/Core/Portable/Completion/CompletionHelper.cs

```diff
@@ -0,0 +1,525 @@
+                if (match != null)
+                {
+                    return match;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Apply any culture-specific quirks to the given text for the purposes of pattern matching.
+        /// For example, in the Turkish locale, capital 'i's should be treated specially in Visual Basic.
+        /// </summary>
+        private string GetCultureSpecificQuirks(string candidate)
+        {
+            // TODO: define way to identify this case w/o language specific check
+            if (_language == LanguageNames.VisualBasic)
+            {
+                if (CultureInfo.CurrentCulture.Name == "tr-TR")
+                {
+                    return candidate.Replace('I', 'İ');
+                }
+            }
+
+            return candidate;
+        }
+
+        private readonly object _gate = new object();
+        private readonly Dictionary<string, PatternMatcher> _patternMatcherMap = new Dictionary<string, PatternMatcher>();
+        private readonly Dictionary<string, PatternMatcher> _fallbackPatternMatcherMap = new Dictionary<string, PatternMatcher>();
+        internal static readonly CultureInfo EnUSCultureInfo = new CultureInfo("en-US");
+
+        private PatternMatcher GetPatternMatcher(string value, CultureInfo culture)
+        {
+            lock (_gate)
+            {
+                PatternMatcher patternMatcher;
+                if (!_patternMatcherMap.TryGetValue(value, out patternMatcher))
+                {
+                    patternMatcher = new PatternMatcher(value, culture, verbatimIdentifierPrefixIsWordCharacter: true);
+                    _patternMatcherMap.Add(value, patternMatcher);
+                }
+
+                return patternMatcher;
+            }
+        }
+
+        private PatternMatcher GetFallbackPatternMatcher(string value)
+        {
+            lock (_gate)
+            {
+                PatternMatcher patternMatcher;
+                if (!_fallbackPatternMatcherMap.TryGetValue(value, out patternMatcher))
+                {
+                    patternMatcher = new PatternMatcher(value, EnUSCultureInfo, verbatimIdentifierPrefixIsWordCharacter: true);
+                    _fallbackPatternMatcherMap.Add(value, patternMatcher);
+                }
+
+                return patternMatcher;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if item1 is a better completion item than item2 given the provided filter
+        /// text, or false if it is not better.
+        /// </summary>
+        public bool IsBetterFilterMatch(CompletionItem item1, CompletionItem item2, string filterText, CompletionTrigger trigger, CompletionFilterReason filterReason, ImmutableArray<string> recentItems = default(ImmutableArray<string>))
+        {
+            if (_language == LanguageNames.VisualBasic)
```

Comment:
Smae concern.  How does a language like TS define what is a better filter match?  Do they need us to change this type?